python/ray/tune/suggest/search.py

```diff
@@ -0,0 +1,160 @@
+
+    See `VariantAlgorithm`.
+    """
+
+    def add_experiment(self, experiment):
+        """Adds an Experiment to the search algorithm.
+
+        Arguments:
+            experiment (Experiment): Specification of a single experiment.
+        """
+        raise NotImplementedError
+
+    def next_trial(self):
+        """Provides Trial objects to be queued into the TrialRunner.
+
+        Returns:
+            Trial|None: If SearchAlgorithm cannot be queried at
+                a certain time (i.e. due to constrained concurrency), this will
+                return None. Otherwise, return a trial.
+        """
+        raise NotImplementedError
+
+    def on_trial_result(self, trial_id, result):
+        """Called on each intermediate result returned by a trial.
+
+        This will only be called when the trial is in the RUNNING state.
+
+        Arguments:
+            trial_id: Identifier for the trial.
+        """
+        pass
+
+    def on_trial_complete(self,
+                          trial_id,
+                          result=None,
+                          error=False,
+                          early_terminated=False):
+        """Notification for the completion of trial.
+
+        Arguments:
+            trial_id: Identifier for the trial.
+            result (TrainingResult): Defaults to None. A TrainingResult will
+                be provided with this notification when the trial is in
+                the RUNNING state AND either completes naturally or
+                by manual termination.
+            error (bool): Defaults to False. True if the trial is in
+                the RUNNING state and errors.
+            early_terminated (bool): Defaults to False. True if the trial
+                is stopped while in PAUSED or PENDING state.
+        """
+        pass
+
+
+class VariantAlgorithm(SearchAlgorithm):
+    """Uses Tune's variant generation for resolving variables.
+
+    Custom search algorithms can extend this class easily by overriding the
+    `try_suggest` method, which will override conflicting fields from
+    the initially generated parameters.
+
+    See `ray.tune.suggest.variant_generator`.
+
+    To track suggestions and their corresponding evaluations, the method
+    `try_suggest` will need to generate a trial_id. This trial_id will
+    be used in subsequent notifications.
+
+    Attributes:
+        PASS (str): Status string for `try_suggest` if VariantAlgorithm
+            currently cannot be queried for parameters (i.e. due to
+            constrained concurrency).
+
+    Example:
+        >>> suggester = VariantAlgorithm()
+        >>> new_parameters, trial_id = suggester.try_suggest()
+        >>> suggester.on_trial_complete(trial_id, result)
+        >>> better_parameters, trial_id2 = suggester.try_suggest()
+    """
+
+    PASS = "PASS"
+
+    def __init__(self):
+        self._generator = []
+
+    def add_experiment(self, experiment):
```

Comment:
This is super weird, I think you can remove it and just pass the experiment config into the constructor of this class instead.

Reply:
That would drop multi-experiment support for `run_experiments`; OK with that.

Reply:
Actually not necessary; I see.